### DISCUSSION TITLE
SSCHA negative frequencies

### DISCUSSION THREAD

**Discussion by matukumilli** [SSCHAcode/python-sscha#67] [Q&A] [answered]
Dear SSCHA developers,

I am facing the below error message:

....
q=  [2.27808477e-19 4.44648911e-19 2.12891636e-03]     (2pi/A)
w(q)=  [-7.25465934e-02 -7.25465932e-02  1.95568180e+00  2.27781659e+01
  2.96765253e+01  2.96765253e+01  3.05479221e+01  4.35835090e+01
  4.35835090e+01  4.42578840e+01  4.42578840e+01  7.92147696e+01
  8.62852524e+01  8.62852524e+01  9.08039457e+01  9.50114310e+01
  1.42969243e+02  1.44039177e+02  1.51649822e+02  1.51649822e+02
  1.69078933e+02  1.74300007e+02  1.74300007e+02  2.03209466e+02]   (cm-1)
Cannot continue with SSCHA negative frequencies

As evident from the below trends, sscha minimisation reached minimum. With the 7th (final) population, it reached the convergence criteria much well before the KL value. Negative freq appearing despite all this.
![plot](https://user-images.githubusercontent.com/4202292/153698333-faac4aa2-07ef-413c-8a1a-c02842e01cf9.png)

I think " tensor3.Apply_ASR() " is somehow not setting the -ve frequencies of acoustic modes to ZERO in sscha dynamical matrix :sscha_dyn_pop7_1 :

     q = (    0.000000000000   0.000000000000   0.000000000000 )

***************************************************************************
   freq (    1) =    -0.00000013 [THz] =    -0.00000427 [cm-1]
   freq (    2) =    -0.00000004 [THz] =    -0.00000134 [cm-1]
   freq (    3) =     0.00000004 [THz] =     0.00000129 [cm-1]

I'd be grateful if you could suggest a way to avoid this error.

Best regards,
Prasad Matukumilli

**Comment by mesonepigreco** [comment, marked as answer]
Dear Prasad Matukumilli,

Which function is raising the error? It is possible that the acoustic frequencies at gamma are not exactly zero due to the error in the diagonalization. However, the code should discard acoustic frequencies anyway in the calculation. If you can be a bit more specific on your input file and exactly which part of the input raises the error, we could try to help you, or fix the code if it is a bug. 

Moreover, the code is not complaining about the acoustic modes at gamma, but at a different q point.
If you are computing spectral functions, you should give the code the SSCHA dynamical matrices, not the free energy hessian (which is not positive definite).

Bests,
Lorenzo

**Reply by matukumilli** [reply]
Dear Dr. Lorenzo Monacelli,

This error is given by the function CC.Spectral.get_static_correction_along_path() while using SSCHA dynamical matrices. Yes, I understood now that error is corresponding to finite q point. Is it possible that interpolation along the q-path might result -ve freq (very small values)?  My input file is as follows:

dyn = CC.Phonons.Phonons("../dyn_pop7_",4)

supercell = dyn.GetSupercell()
tensor3 = CC.ForceTensor.Tensor3(dyn.structure,
                                dyn.structure.generate_supercell(supercell),
                                supercell)

d3 = np.load("d3_realspace_sym.npy")*2.0
tensor3.SetupFromTensor(d3)

tensor3.Center()     
tensor3.Apply_ASR()  

k_grid=[10, 10, 10]

path = ase.dft.kpoints.bandpath("GXMGZ", dyn.structure.unit_cell, npoints = 200)
qpath = path.cartesian_kpts()

CC.Spectral.get_static_correction_along_path(dyn=dyn,
                                             tensor3=tensor3,
                                             k_grid=k_grid,
                                             q_path = qpath,   
                                             filename_st="v2_v2+d3static_freq-GXMGZ.dat",
                                             T =300.0,
                                             print_dyn = False) 


I have updated my sscha installation on 11/Dec/2021. Please let me know in any other details are needed.

Warm regards,
Prasad

**Reply by mesonepigreco** [reply]
I see,
So the problem is that you have a (small) imaginary frequency of the acoustic mode close to gamma when interpolating the SSCHA dyn.
This is due to an interpolation error that is a consequence of using a too-small supercell for the SSCHA calculation.

You can improve the interpolation using a support harmonic dynamical matrix.
You can compute the harmonic phonons on a bigger mesh than the SSCHA one, and then interpolate only the difference between harmonic and SSCHA phonons, which usually behaves much better

```
# [ Here the previous par tof the input]
harm_dyn = CC.Phonons.Phonons("../harmonic_dyn",4)
harm_dyn_fine_grid = CC.Phonons.Phonons("../harmonic_dyn", NQIRR_FINE_GRID)
new_cell = harm_dyn_fine_grid.GetSupercell()
big_dyn = dyn.Interpolate( dyn.GetSupercell(), new_cell, support_dyn_coarse = harm_dyn, support_dyn_fine = harm_dyn_fine_grid)
big_dyn.ForcePositiveDefinite()
CC.Spectral.get_static_correction_along_path(dyn=big_dyn,
                                            tensor3=tensor3,
                                            k_grid=k_grid,
                                            q_path = qpath,
                                            filename_st="v2_v2+d3static_freq-GXMGZ.dat",
                                            T =300.0,
                                            print_dyn = False)
```

Here NQIRR_FINE_GRID is the number of irreducible points in the new supercell. This should definitively prevent imaginary frequencies to occur. If not, try to reduce a bit k_grid and npoints.

**Reply by matukumilli** [reply]
Thank you, Dr. Lorenzo! With harmonic phonons on finer grid, I got the correct interpolation.

Regards,
Prasad

**Reply by mesonepigreco** [reply]
Hi, glad this solved the issue!

**Reply by matukumilli** [reply]
Hi, 

When I try this interpolation way to avoid -ve freq error, for tetragonal SrTiO3, I getting this below error:

....
...
 Iter #  128  ====
 
         Sum on 3rd=        0.844989E-12    Imp. ASR on 3rd:  => delta FC=        0.844988E-12
                                            Imp. permut sym:  => delta FC=        0.800810E-12
 
  
 * Convergence reached within threshold:        0.100000E-11
  
 * Total FC relative variation:        0.128277E+00
 
 Time elapsed for imposing ASR: 186.2998547554016 s
 
 ============================================================ 
WORKING ON: [0. 0. 0.]
/home/waghmare/.local/lib/python3.5/site-packages/cellconstructor/Phonons.py:3661: ComplexWarning: Casting complex values to real discards the imaginary part
  QE_fc[:,:, i, j] = starting_fc[3*i : 3*(i+1), 3*j : 3*(j+1)]
 SUPERCELL SIZE:           2           2           2
 NAT:          10
 The unit cell:
   5.5080868904697313        7.3440103850181005E-003   0.0000000000000000     
   3.0484109872987837E-002   5.5080073591324155        0.0000000000000000     
  -2.7692854120418908       -2.7576755993832940        3.8814121235103820     
 Total weight:   7.6666666666666670     
 NR1,2,3 =            2           2           2

 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
     Error in routine frc_blk (1):
     wrong total_weight
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

     stopping ...
STOP 1

I couldn't able to figure out the reason for this. Need your help.

Regards,
Prasad
